--- build.rs
+++ build.rs
@@ -134,7 +134,11 @@ fn build_nasm_files() {
 
   if let Some((hash, hash_path)) =
     hash_changed(asm_files, &out_dir, &dest_path)
   {
-    let obj = nasm_rs::Build::new()
+    let mut nasm_build = nasm_rs::Build::new();
+    if let Ok(nasm) = env::var("NASM") {
+      nasm_build.nasm(nasm);
+    }
+    let obj = nasm_build
       .min_version(2, 15, 0)
       .include(&out_dir)
       .include("src")
