commit 2bbe2e3007dbb9a69b207ebbe1c12cbd5d92a8e3
Author: David Zbarsky <dzbarsky@gmail.com>
Date:   Mon Dec 22 19:56:34 2025 -0500

    [windows] Add dllcrt to crt dir

diff --git a/runtimes/BUILD.bazel b/runtimes/BUILD.bazel
index e800f2d..f2c657d 100644
--- a/runtimes/BUILD.bazel
+++ b/runtimes/BUILD.bazel
@@ -105,6 +105,7 @@ copy_to_directory(
         "//runtimes/mingw:crtbegin.o",
         "//runtimes/mingw:crtend.o",
         "//runtimes/mingw:crt2.o",
+        "//runtimes/mingw:dllcrt2.o",
     ],
     target_compatible_with = [
         "@platforms//os:windows",
diff --git a/runtimes/mingw/BUILD.bazel b/runtimes/mingw/BUILD.bazel
index abd5799..dcd63b1 100644
--- a/runtimes/mingw/BUILD.bazel
+++ b/runtimes/mingw/BUILD.bazel
@@ -263,6 +263,18 @@ crt_object(
     visibility = ["//visibility:public"],
 )
 
+crt_object(
+    name = "mingw_dllcrt2",
+    out = "dllcrt2.o",
+    srcs = ["@mingw//:mingw-w64-crt/crt/crtdll.c"],
+    implementation_deps = [
+        "@mingw//:mingw_headers",
+        "@mingw//:mingw_crt_headers",
+    ],
+    copts = MINGW_CRT_OBJECT_COPTS,
+    visibility = ["//visibility:public"],
+)
+
 copy_file(
     name = "version_lib",
     src = select({
diff --git a/runtimes/mingw/BUILD.tpl b/runtimes/mingw/BUILD.tpl
index 8fe2a53..6e490e2 100644
--- a/runtimes/mingw/BUILD.tpl
+++ b/runtimes/mingw/BUILD.tpl
@@ -25,6 +25,7 @@ package(default_visibility = ["//visibility:public"])
 
 exports_files([
     "mingw-w64-crt/crt/crtbegin.c",
+    "mingw-w64-crt/crt/crtdll.c",
     "mingw-w64-crt/crt/crtexe.c",
     "mingw-w64-crt/crt/crtend.c",
 ])
